### PR TITLE
RFC, WIP : Support for Time Dependent Equations

### PR DESCRIPTION
**Pull request by amitjamadagni**
This is an initial attempt at supporting time dependent equations in QuDynamics. Though, currently the support is for `QuSchrodingerEq` type it will be extended to other similar types. The usage is as follows : 

``` julia
# requires `complex` from QuBase, currently not supported by the base.
julia> using QuBase

julia> using QuDynamics

julia> seq = QuSchrodingerEqTD(t->t*coeffs(sigmax)+t/2*coeffs(sigmay)+t/3*coeffs(sigmaz))
QuDynamics.QuSchrodingerEqTD((anonymous function))

julia> seq.hamiltonian(2)
2x2 Array{Complex{Float64},2}:
 0.666667+0.0im        2.0-1.0im
      2.0+1.0im  -0.666667+0.0im

julia> setd = QuStateEvolution(seq, complex(statevec(1, FiniteBasis(2))), 0.:0.1:2*π, QuExpokit())
QuDynamics.QuStateEvolution{QuDynamics.QuExpokit,QuBase.QuArray{QuBase.FiniteBasis{QuBase.Orthonormal},Complex{Float64},1,Array{Complex{Float64},1}},QuDynamics.QuSchrodingerEqTD}(QuDynamics.QuSchrodingerEqTD((anonymous function)),2-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[1.0 + 0.0im,0.0 + 0.0im],0.0:0.1:6.2,QuDynamics.QuExpokit(Dict{Symbol,Any}()))

julia> for (t, psi) in setd
           println(t, psi)
       end
0.12-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[0.9999319452163673 - 0.003333257716563999im,0.0049998865748459975 - 0.009999773149691995im]
0.22-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[0.9993875625234887 - 0.009997958458381771im,0.014996937687572657 - 0.029993875375145314im]
0.300000000000000042-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[0.9975510002532798 - 0.019983670667866514im,0.029975506001799768 - 0.059951012003599535im]
0.42-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[0.9932021602074401 - 0.033257767494471145im,0.049886651241706714 - 0.09977330248341343im]
0.52-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[0.9847265389049336 - 0.04974518216959884im,0.07461777325439828 - 0.14923554650879656im]
0.60000000000000012-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[0.9701373249726354 - 0.06930180708244918im,0.10395271062367377 - 0.20790542124734754im]
0.70000000000000012-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[0.9471172291701745 - 0.09168221700537858im,0.13752332550806787 - 0.27504665101613573im]
.
.
.
5.02-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[-0.6726861118655973 - 0.21140799029232685im,0.31711198543849006 - 0.6342239708769803im]
5.10000000000000052-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[-0.971820096386157 - 0.06734965130888733im,0.10102447696333089 - 0.2020489539266619im]
5.22-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[-0.9327954328048941 + 0.10297322042928472im,-0.15445983064392702 + 0.3089196612878539im]
5.3000000000000012-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[-0.5511648169336071 + 0.23839897010025188im,-0.3575984551503776 + 0.7151969103007553im]
5.42-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[0.04622392867070896 + 0.2854088865539027im,-0.4281133298308537 + 0.8562266596617079im]
5.52-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[0.6349216268345304 + 0.22073615754372197im,-0.33110423631558283 + 0.6622084726311661im]
5.60000000000000052-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[0.9737675389036275 + 0.06501289051305957im,-0.09751933576958946 + 0.1950386715391791im]
5.72-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[0.906682813602937 - 0.120518023215452im,0.18077703482317775 - 0.3615540696463554im]
5.8000000000000012-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[0.44276993597104014 - 0.2561815019529585im,0.38427225292943734 - 0.7685445058588747im]
5.92-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[-0.22763453703350778 - 0.2782133255145054im,0.41731998827175754 - 0.8346399765435155im]
6.02-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[-0.8014093051320135 - 0.17089037556587786im,0.2563355633488165 - 0.5126711266976336im]
6.10000000000000052-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[-0.9975146295768043 + 0.020131331616999027im,-0.030196997425498362 + 0.06039399485099639im]
6.22-element QuVector in QuBase.FiniteBasis{QuBase.Orthonormal}:
...coefficients: Array{Complex{Float64},1}
Complex{Float64}[-0.7011042299260944 + 0.2037310870209901im,-0.3055966305314848 + 0.6111932610629696im]
```

@acroy any thoughts on this will be helpful. Thanks :-)

### CODE REVIEW

File: src/quequations.jl [hunk 1]
```diff
@@ -17,6 +17,11 @@ immutable QuSchrodingerEq{H<:QuBase.AbstractQuMatrix} <: QuEquation{1}
     QuSchrodingerEq(hamiltonian) = new(hamiltonian)
 end
 
+immutable QuSchrodingerEqTD <: QuEquation{1}
```
Comment (acroy) [inline src/quequations.jl]: I think this should be `<:QuEquation{1}` since there is no caching involved.

**Comment by amitjamadagni** [comment]
@acroy using the recent commit, the following are the results for time-dependent with and without collapse operators. 

Without collpase operators : https://github.com/amitjamadagni/Notes-Notebooks/blob/master/Notebooks/Time%20Dependence%20without%20collapse%20operators.ipynb

With collapse operators https://github.com/amitjamadagni/Notes-Notebooks/blob/master/Notebooks/Time%20Dependence%20with%20collapse%20operators.ipynb

This in comparison with the results from QuTiP time dependent version seem to match. Here are the results from the QuTiP version.

QuTiP version : https://github.com/amitjamadagni/Notes-Notebooks/blob/master/Notebooks-python/Time%20Dependence.ipynb

Thoughts on this would be helpful !

**Comment by amitjamadagni** [comment]
``` julia
julia> using QuBase

julia> using QuDynamics

julia> immutable QuLindbladMasterEqTD <: QuEquation{1} end

julia> function h_td(H,t)
           return expm(H*t)
       end
h_td (generic function with 1 method)

julia> function cops_td(cops, t)
           cop = []
           for op in cops
               push!(cop, expm(op*t))
           end
           return cop
       end
cops_td (generic function with 1 method)

julia> function QuDynamics.operator(eq::QuLindbladMasterEqTD, t)
           QuDynamics.lindblad_op(h_td(sigmax, t), cops_td([lowerop(2)], t))
       end
operator (generic function with 9 methods)

julia> qprop = QuPropagator(QuLindbladMasterEqTD(), statevec(1, FiniteBasis(2))*statevec(1, FiniteBasis(2))', 0.:0.1:2*π, QuExpmV())
Summarizing the system :
Time steps used : 0.0:0.1:6.2
Solver used : QuDynamics.QuExpmV(Dict{Symbol,Any}())
```

`@show` does not work as expected for new types.

**Comment by amitjamadagni** [comment]
@acroy the failure is because of `normalize` issue in the QuBase. I have used `Compat` to fix the problem as of now. The PR in QuBase regarding the same is https://github.com/JuliaQuantum/QuBase.jl/pull/47, once that is merged we can merge this as well.

**Comment by amitjamadagni** [comment]
Travis build pass. Merging this !

**Pull request merged**